src/main/java/org/embulk/filter/column/JsonColumn.java

```diff
@@ -39,33 +44,45 @@ public JsonColumn(String path, Type type, Value defaultValue)
 
     public JsonColumn(String path, Type type, Value defaultValue, String src)
     {
-        this.path = path;
+        CompiledPath compiledPath = PathCompiler.compile(path);
+        CompiledPath compiledSrc = src == null ? compiledPath : PathCompiler.compile(src);
+        this.path = compiledPath.toString();
         this.type = type;
         this.defaultValue = (defaultValue == null ? ValueFactory.newNil() : defaultValue);
-        this.src = (src == null ? path : src);
+        this.src = compiledSrc.toString();
 
         this.pathValue = ValueFactory.newString(path);
-        this.parentPath = parentPath(path);
-        this.baseName = baseName(path);
-        if (this.baseName.equals("[*]")) {
+        this.parentPath = compiledPath.getParentPath();
+        this.tailPath = compiledPath.getTailPath();
+        if (this.tailPath.equals("[*]")) {
             throw new ConfigException(String.format("%s wrongly ends with [*], perhaps you can remove the [*]", path));
         }
-        this.baseIndex = baseIndex(path);
+        this.tailIndex = compiledPath.tailIndex();
         this.parentPathValue = ValueFactory.newString(parentPath);
-        this.baseNameValue = ValueFactory.newString(baseName);
+        String baseName = getPropertyOrTailName(compiledPath, this.tailPath);
+        this.tailNameValue = ValueFactory.newString(baseName);
 
         this.srcValue = ValueFactory.newString(this.src);
-        this.srcParentPath = parentPath(this.src);
-        this.srcBaseName = baseName(this.src);
-        this.srcBaseIndex = baseIndex(this.src);
+        this.srcParentPath = compiledSrc.getParentPath();
+        this.srcTailName = compiledSrc.getTailPath();
+        this.srcTailIndex = compiledSrc.tailIndex();
         this.srcParentPathValue = ValueFactory.newString(this.srcParentPath);
-        this.srcBaseNameValue = ValueFactory.newString(this.srcBaseName);
+        String srcTailName = getPropertyOrTailName(compiledSrc, this.srcTailName);
+        this.srcTailNameValue = ValueFactory.newString(srcTailName);
 
         if (! srcParentPath.equals(parentPath)) {
             throw new ConfigException(String.format("The branch (parent path) of src \"%s\" must be same with of name \"%s\" yet", src, path));
         }
     }
 
```

Comment:
If it is possible, I may want to change the behavior as

```
$['foo'][0] or $.foo[0] => NULL
```

but, if it makes things difficult, it is okay to keep.